data/src/main/java/com/microsoft/azure/kusto/data/auth/CloudDependentTokenProviderBase.java

```diff
@@ -42,9 +52,22 @@ protected void initializeWithCloudInfo(CloudInfo cloudInfo) throws DataClientExc
 
     @Override
     public String acquireAccessToken() throws DataServiceException, DataClientException {
+        return acquireAccessTokenInner();
+    }
+
+    @Override
+    protected String acquireAccessTokenInner() throws DataServiceException, DataClientException {
         initialize();
-        return acquireAccessTokenImpl();
+        Map<String, String> attributes = cloudInfo != null ? new HashMap<>(Map.of("resource", cloudInfo.getKustoServiceResourceId())) : null;
```

Comment:
Why do we need these three levels of nesting.
I'd have expected acquireAccessToken() not to be overridable and do what acquireAccessTokenInner() does.
acquireAccessTokenImpl() should be abstract and overrideable.

Reply:
That's what we discussed the other day, the nesting makes it so that cloudInfo providers can have their attributes before the run.